File: shadow/test/dispatch.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { buildDispatchInputs, extractRunId, classifyRunState } from '../src/core/dispatch.ts';

describe('buildDispatchInputs', () => {
  it('maps the shadow context to the receiver workflow_dispatch inputs (all strings)', () => {
    assert.deepEqual(
      buildDispatchInputs({
        workflowsRepo: 'stefanpenner-cs/reusable-workflows',
        workflowsRef: 'deadbeef',
        consumerRepo: 'o/r',
        consumerRef: 'main',
        workflowsPr: 7,
        branch: 'shadow/pr-7-o-r',
      }),
      {
        workflows_repo: 'stefanpenner-cs/reusable-workflows',
        workflows_ref: 'deadbeef',
        consumer_repo: 'o/r',
        consumer_ref: 'main',
        workflows_pr: '7',
        branch: 'shadow/pr-7-o-r',
      },
    );
  });
});

describe('extractRunId', () => {
  // Captured shape of the return_run_details response (REST: workflow_run_id / run_url / html_url).
  const response = {
    workflow_run_id: 1234567890,
    run_url: 'https://api.github.com/repos/o/h/actions/runs/1234567890',
    html_url: 'https://github.com/o/h/actions/runs/1234567890',
  };

  it('reads workflow_run_id', () => {
    assert.equal(extractRunId(response), 1234567890);
  });

  it('throws when no run id is present', () => {
    assert.throws(() => extractRunId({}));
  });

  it('throws when workflow_run_id is not a number', () => {
    assert.throws(() => extractRunId({ workflow_run_id: 'nope' }));
  });
});

describe('classifyRunState', () => {
  it('is pending while not completed', () => {
    assert.equal(classifyRunState('queued', null), 'pending');
    assert.equal(classifyRunState('in_progress', null), 'pending');
  });

  it('is success only when completed + success', () => {
    assert.equal(classifyRunState('completed', 'success'), 'success');
  });

  it('is failure for any non-success conclusion', () => {
    assert.equal(classifyRunState('completed', 'failure'), 'failure');
    assert.equal(classifyRunState('completed', 'cancelled'), 'failure');
    assert.equal(classifyRunState('completed', null), 'failure');
  });
});

File: internal/shadow/core/dispatch_test.go
package core

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestBuildDispatchInputs(t *testing.T) {
	got := BuildDispatchInputs(ShadowContext{
		WorkflowsRepo: "stefanpenner-cs/reusable-workflows",
		WorkflowsRef:  "deadbeef",
		ConsumerRepo:  "o/r",
		ConsumerRef:   "main",
		WorkflowsPR:   7,
		Branch:        "shadow/pr-7-o-r",
	})
	assert.Equal(t, map[string]string{
		"workflows_repo": "stefanpenner-cs/reusable-workflows",
		"workflows_ref":  "deadbeef",
		"consumer_repo":  "o/r",
		"consumer_ref":   "main",
		"workflows_pr":   "7",
		"branch":         "shadow/pr-7-o-r",
	}, got)
}

func TestExtractRunID(t *testing.T) {
	id, err := ExtractRunID(map[string]any{"workflow_run_id": float64(1234567890)})
	require.NoError(t, err)
	assert.Equal(t, 1234567890, id)

	_, err = ExtractRunID(map[string]any{})
	assert.Error(t, err)
	_, err = ExtractRunID(map[string]any{"workflow_run_id": "nope"})
	assert.Error(t, err)
}

func TestClassifyRunState(t *testing.T) {
	assert.Equal(t, RunPending, ClassifyRunState("queued", ""))
	assert.Equal(t, RunPending, ClassifyRunState("in_progress", ""))
	assert.Equal(t, RunSuccess, ClassifyRunState("completed", "success"))
	assert.Equal(t, RunFailure, ClassifyRunState("completed", "failure"))
	assert.Equal(t, RunFailure, ClassifyRunState("completed", "cancelled"))
	assert.Equal(t, RunFailure, ClassifyRunState("completed", ""))
}
